File: codigo-verificador-base10.swift
/*
 * Crear y validar números de longitud fija con algoritmo de Luhn
 * @ToRo 2017
 * https://tar.mx
 * v1.1
 */
import Foundation
//MARK: padding izquierdo
extension String {
    func left(total:Int, cadena: String) -> String {
        let pad = total - self.characters.count;
        return pad < 1 ? self : "".padding(toLength: pad, withPad: cadena, startingAt: 0)+self
    }
}
//MARK: crear dígito
func digito(cadena:String) -> Int {
    var sum = [Int](); var a = 2 ; var total = 0;
    //multiplicamos
    for i in cadena.characters {
        let n = Int(String(i))
        a = (a < 1) ? 2 : a //primer caracter *1, luego *2 y así.
        sum.append(n!*a)    // añadimos al arreglo
        a = a-1
    }
    sum.reverse()
    //print(sum)
    //sumamos
    for numero in sum {
        for chars in String(Int(numero)).characters {
            let suma = Int(String(chars))
            total += suma!
        }
    }
    total %= 10             //base10
    return (total != 0) ? 10-total : total
}
//MARK: ejemplo
let longitud = 5;           //longitud de relleno de cada número
let prefijo  = "01";        //prefijo número
let numeros  = [37,92,10,20,324]; //números de ejemplo
for numero in numeros {
    let original = String(numero)
    let relleno = prefijo + original.left(total: longitud, cadena: "0")
    let digitov = digito(cadena: relleno)   //aquí obtiene el prefijo
    let codigo = relleno + String(Int(digitov))
    print("Final: \(codigo) Inicial: \(original) - Relleno \(relleno) Dígito verificador \(digitov)")
}
//EOF

File: codigo-verificador-base10.php
<?php
   /*
   * https://tar.mx/archivo/2016/crear-y-validar-numeros-de-longitud-fija-con-algoritmo-de-luhn.html
   * @ToRo 2016
   * https://tar.mx/tema/codigo-de-barras.html
   */
   // genera dígito verificador según Algoritmo de Luhn
   // https://es.wikipedia.org/wiki/Algoritmo_de_Luhn
   function digito($digito) {
      $a=2;
      $sum = [];
      for($i=strlen($digito)-1;$i>=0;$i--) {
         $d =$digito[$i];
         //si es el primero es *2, de lo contrario *1;
         if($a<1) $a=2;
         $sum[] = $d*$a;
         $a--;
      }
      //ahora sumamos
      $total = 0;
      foreach($sum AS $d) {
         if(strlen($d)==1) $total += $d;
         else {
            $da = str_split($d);
            foreach($da AS $one) { $total += $one; }
         }
      }
      $total %= 10;
      if($total != 0) $total = 10-$total;
      return $total;
   }
   // verifica que un número sea válido
   function verifica($numero) {
      $numero_checksum = '';
      foreach (str_split(strrev((string) $numero)) as $i => $d) {
         $numero_checksum .= $i %2 !== 0 ? $d * 2 : $d;
      }
      return array_sum(str_split($numero_checksum)) % 10 === 0;
   }
   //
   //
   $longitud=5;
   $numeros = [1,2,3,435,50230]; // id de mis clientes, podría ser obtenido de una DB
   $prefijo = "01"; //producto 01, se tiene de 1 a 99 tipo de tarjetas o productos
   foreach($numeros AS $k) {
      $numero = $prefijo;
      $numero .= str_pad($k,$longitud,0,STR_PAD_LEFT); // 1 = 000001, etc.
      $digito = digito($numero); //del prefijo + número
      $tarjeta = $numero.$digito;
      echo "Prefijo + ID : $numero, DV: ".$digito.", tarjeta: ".$tarjeta."\n";
   }
   //
   /*
   // ejemplo para generar la tarjeta en código de barras
   $tarjeta = "01000017"; //(si nuestra tarjeta fueran esos 8 códigos)
   $tmpf = "/tmp/".$tarjeta.".ps"; //archivo temporal
   $tmpi = $tarjeta.".png"; //archivo final
   $cmd = "barcode -e 128 -E -b $tarjeta -o $tmpf";
   $cmd = `$cmd`;
   $cmd = "convert -density 300 $tmpf $tmpi";
   $cmd = `$cmd`;
   unlink($tmpf);
   */
   //
   // verificamos el número según el algoritmo de Luhn:
   $tarjeta = "01000017";
   echo "Tarjeta $tarjeta es válida: ";
   echo verifica($tarjeta);
   echo "\n";
